File: UnzipAndOpen.java
/*
 * Inspired by: Josh Long
 * 
*/

import java.io.File;
import java.util.Set;

public class UnzipAndOpen {
    public static void main(String[] args) throws Exception {
        // Ensure an argument is passed
        if (args.length == 0) {
            System.out.println("You must specify a zip file to open.");
            return;
        }

        var zipFile = new File(args[0]);

        // Check if the file exists and is a .zip file
        if (!zipFile.exists() || !zipFile.getName().endsWith(".zip")) {
            System.out.println("Zip file not found or invalid file type.");
            return;
        }

        var zipFileAbsolutePath = zipFile.getAbsolutePath();
        var folder = new File(zipFileAbsolutePath.substring(0, zipFileAbsolutePath.lastIndexOf(".")));

        try {
            // Unzipping the file
            new ProcessBuilder().command("unzip", "-a", zipFileAbsolutePath).inheritIO().start().waitFor();
        } catch (Exception e) {
            System.out.println("Error during unzip process: " + e.getMessage());
        }

        // Check for build files (Gradle Groovy, Kotlin, or Maven POM)
        for (var k : Set.of("build.gradle", "build.gradle.kts", "pom.xml")) {
            var buildFile = new File(folder, k);
            if (buildFile.exists()) {
                try {
                    // Open in IntelliJ IDEA
                    new ProcessBuilder().command("idea", buildFile.getAbsolutePath()).inheritIO().start().waitFor();
                } catch (Exception e) {
                    System.out.println("Error opening file in IDEA: " + e.getMessage());
                }
            }
        }
    }
}

File: UnzipAndOpen.py
'''
Inspired by: Josh Long
'''

import subprocess
import sys
import os

def run(c):
    print('running: %s' % c)
    return subprocess.check_call(c, shell=True)

if __name__ == '__main__':
    # Check for .zip argument
    assert len(sys.argv) > 1 and sys.argv[1].endswith('.zip'), 'You must specify a .zip file to open.'
    
    zip_file = os.path.abspath(sys.argv[1])
    assert os.path.exists(zip_file), 'Zip file not found.'
    
    folder_name = os.path.splitext(os.path.basename(zip_file))[0]  # Correct way to get the folder name
    try:
        run(f'unzip -a {zip_file}')  # Use f-string for better readability
    except subprocess.CalledProcessError:
        print("Non-zero exit code, but proceeding anyway...")

    # Define build files for Gradle (Groovy and Kotlin) and Maven (pom.xml)
    gradle_build_groovy = os.path.join(folder_name, 'build.gradle')
    gradle_build_kotlin = os.path.join(folder_name, 'build.gradle.kts')
    mvn_pom = os.path.join(folder_name, "pom.xml")
    
    success = False
    build_files = [gradle_build_groovy, gradle_build_kotlin, mvn_pom]

    # Check for the existence of any build file and open it in IntelliJ IDEA
    for fn in build_files:
        if os.path.exists(fn):
            run(f'idea {fn}')
            success = True
    
    # If no build file is found, raise an error
    assert success, 'Valid build file (one of: %s) not found' % ', '.join(build_files)
    print("....Done")
